File: 백준/Silver/1991. 트리 순회/트리 순회.py
n = int(input())

tree = {}

for _ in range(n):
    parent, left, right = input().split()
    tree[parent] = []
    tree[parent].append(left)
    tree[parent].append(right)


def pre_order(root):
    global tree

    L, R = 0, 1

    print(root, end='')
    
    if tree[root][L] != '.':
        pre_order(tree[root][L])
        
    if tree[root][R] != '.':
        pre_order(tree[root][R])


def in_order(root):
    global tree

    L, R = 0, 1

    
    if tree[root][L] != '.':
        in_order(tree[root][L])

    print(root, end='')
        
    if tree[root][R] != '.':
        in_order(tree[root][R])


def post_order(root):
    global tree

    L, R = 0, 1

    if tree[root][L] != '.':
        post_order(tree[root][L])
    
    if tree[root][R] != '.':
        post_order(tree[root][R])

    print(root, end='')
        

pre_order('A')
print()
in_order('A')
print()
post_order('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {

    static int n;
    static int[][] tree;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());

        tree = new int[n][2];
        for (int i = 0; i < n; i++) {
            StringTokenizer tokenizer = new StringTokenizer(br.readLine());

            char parent = tokenizer.nextToken().charAt(0);
            char left = tokenizer.nextToken().charAt(0);
            char right = tokenizer.nextToken().charAt(0);

            tree[parent - 'A'][0] = left == '.' ? -1 : left - 'A';
            tree[parent - 'A'][1] = right == '.' ? -1 : right - 'A';
        }

        StringBuilder answer = new StringBuilder();
        preOrder(answer, 0);
        answer.append("\n");
        inOrder(answer, 0);
        answer.append("\n");
        postOrder(answer, 0);

        System.out.println(answer);
        br.close();
    }

    static void preOrder(StringBuilder path, int parent) {
        path.append((char) (parent + 'A'));
        if (tree[parent][0] != -1)
            preOrder(path, tree[parent][0]);

        if (tree[parent][1] != -1)
            preOrder(path, tree[parent][1]);
    }

    static void inOrder(StringBuilder path, int parent) {
        if (tree[parent][0] != -1)
            inOrder(path, tree[parent][0]);
        path.append((char) (parent + 'A'));
        if (tree[parent][1] != -1)
            inOrder(path, tree[parent][1]);
    }

    static void postOrder(StringBuilder path, int parent) {
        if (tree[parent][0] != -1)
            postOrder(path, tree[parent][0]);

        if (tree[parent][1] != -1)
            postOrder(path, tree[parent][1]);

        path.append((char) (parent + 'A'));
    }
}